src/db.c

```diff
@@ -1070,6 +1070,107 @@ void moveCommand(client *c) {
     addReply(c,shared.cone);
 }
 
+void copyCommand(client *c) {
+    robj *o;
+    redisDb *src, *dst;
+    int srcid;
+    long long dbid, expire;
+    int j, replace = 0, delete = 0;
+
+    /* Obtain source and target DB pointers 
+     * Default target DB is the same as the source DB 
+     * Parse the REPLACE option and targetDB option. */
+    src = c->db;
+    srcid = c->db->id;
+    dbid = c->db->id;
+    for (j = 3; j < c->argc; j++) {
+        int additional = c->argc - j - 1;
+        if (!strcasecmp(c->argv[j]->ptr,"replace")) {
+            replace = 1;
+        } else if (!strcasecmp(c->argv[j]->ptr, "db") && additional >= 1) {
+            if (getLongLongFromObject(c->argv[j+1], &dbid) == C_ERR ||
+                dbid < INT_MIN || dbid > INT_MAX ||
+                selectDb(c, dbid) == C_ERR)
+            {
+                addReplyError(c,"invalid DB index");
+                return;
+            }
+            j++; /* Consume additional arg. */
+        } else {
+            addReply(c, shared.syntaxerr);
+            return;
+        }
+    }
+    dst = c->db;
+    selectDb(c,srcid); /* Back to the source DB */
+
+    if ((server.cluster_enabled == 1) && (srcid != 0 || dbid != 0)) {
+        addReplyError(c,"Copying to another database is not allowed in cluster mode");
+        return;
+    }
+
+    /* If the user select the same DB as
+     * the source DB and using newkey as the same key
+     * it is probably an error. */
+    robj *key = c->argv[1];
+    robj *newkey = c->argv[2];
+    if (src == dst && (sdscmp(key->ptr, newkey->ptr) == 0)) {
+        addReply(c,shared.sameobjecterr);
+        return;
+    }
+
+    /* Check if the element exists and get a reference */
+    o = lookupKeyWrite(c->db, key);
+    if (!o) {
+        addReply(c,shared.czero);
+        return;
+    }
+    expire = getExpire(c->db,key);
+
+    /* Return zero if the key already exists in the target DB. 
+     * If REPLACE option is selected, delete newkey from targetDB. */
+    if (lookupKeyWrite(dst,newkey) != NULL) {
+        if (replace) {
+            delete = 1;
+        } else {
+            addReply(c,shared.czero);
+            return;
+        }
+    }
+
+    /* Duplicate object according to object's type. */
+    robj *newobj;
+    switch(o->type) {
+        case OBJ_STRING: newobj = dupStringObject(o); break;
+        case OBJ_LIST: newobj = dupListObject(o); break;
+        case OBJ_SET: newobj = dupSetObject(o); break;
+        case OBJ_ZSET: newobj = dupZsetObject(o); break;
+        case OBJ_HASH: newobj = dupHashObject(o); break;
+        case OBJ_STREAM: newobj = dupStreamObject(o); break;
+        case OBJ_MODULE:
+            addReplyError(c, "Copying module type object is not supported");
+            return;
+        default: {
+            addReplyError(c, "unknown type object");
+            return;
+        };
+    }
+
+    if (delete) {
+        dbDelete(dst,newkey);
+    }
```

Comment:
Ref: the `REPLACE` logic in `RESTORE`

```suggestion
    if (delete && dbDelete(dst,newkey)) {
        notifyKeyspaceEvent(NOTIFY_GENERIC,"del",c->argv[2],dst->id);
    }
```

Reply:
@itamarhaber are you sure?
the "del" notification in restore is only in case the new key is expired, and restore deleted the old key.
it's something i recently copied from EXPIREAT (when ttl is in the past).

in the case restore succeeds it only does:
```
    notifyKeyspaceEvent(NOTIFY_GENERIC,"restore",key,c->db->id);
```

taking another example, SINTERSTORE it does:
```
   notifyKeyspaceEvent(NOTIFY_SET,"sinterstore", dstkey,c->db->id);
```
or (only in case a key was deleted, and the stored key is empty):
```
   notifyKeyspaceEvent(NOTIFY_GENERIC,"del", dstkey,c->db->id);
```

so i think in this case we just want the "copy_to" notification, not the "del" one

Reply:
@oranagra You are of course correct - the `del` isn't needed here.